server/src/main/java/io/crate/role/Roles.java

```diff
@@ -78,17 +78,27 @@ default Role getUser(String userName) {
     }
 
     /**
-     * Finds a user by given predicate
+     * Finds a user by given predicate.
+     * In case of matching 2 users (by iss/username and one by username),
+     * match by username takes precedence.
      */
     @Nullable
     default Role findUser(Predicate<Role> predicate) {
+        Role matchWithJWT = null;
```

Comment:
I don't like that the JWT related code spreads now into the generic `Roles` interface. What about sorting the roles list by the availability of the `jwtProperties` instead?

Reply:
Removing any mention of JWT here sounds good.

> availability of the jwtProperties instead?

I think even this shouldn't be done here then, right?

I would change signature to return List and then either:
- caller components would pick up one out of 2
- caller components would use smth like `Credentials.getSingleMatch(List<Roles)` which would contain JWT specific logic. wdyt?

Reply:
If the list is sorted by default based on the availability of jwtProperties, we can just return the first match.

Reply:
And yes, probably it should be sorted at the implementation, thus `RolesService`.

Reply:
But it's not super important, you can also leave it like that for now. I think we should first refactor the `Roles` implementations anyhow, e.g. merge `RolesService` and `RolesManagerService`...

Reply:
+1 on not adding JWT specific logic to the `findUser`.
It also doesn't seem right to me to sort the roles based on `jwtProperties` presence.

If ordering is necessary, I think the consumer of this API could just retrieve all roles, sort them, and then do the match itself.
Or use the logic as it is right now but on the callsite.

Reply:
After last discussion, we cannot have multiple matches and `findUser` again doesn't need any JWT specific logic. 

See Roles https://github.com/crate/crate/pull/16763/commits/12622b548b796184c71ed32ad698996b1a42bbae